mllib/src/test/scala/org/apache/spark/ml/feature/TargetEncoderSuite.scala

```diff
@@ -0,0 +1,538 @@
+
+    val data_null = Row(2.toShort, 3, null, 90.0, 60.0, 50.0, 90.0, 57.5, 50.0, 70.0)
+
+    val df_null = spark
+      .createDataFrame(sc.parallelize(data_continuous.dropRight(1) :+ data_null), schema)
+
+    val encoder = new TargetEncoder()
+      .setLabelCol("label")
+      .setTargetType(TargetEncoder.TARGET_CONTINUOUS)
+      .setInputCols(Array("input1", "input2", "input3"))
+      .setOutputCols(Array("output1", "output2", "output3"))
+
+    val model = encoder.fit(df_null)
+
+    val expected_stats = Array(
+      Map(Some(0.0) -> (3.0, 40.0), Some(1.0) -> (3.0, 50.0), Some(2.0) -> (3.0, 60.0),
+        Some(-1.0) -> (9.0, 50.0)),
+      Map(Some(3.0) -> (5.0, 50.0), Some(4.0) -> (4.0, 50.0), Some(-1.0) -> (9.0, 50.0)),
+      HashMap(Some(5.0) -> (3.0, 20.0), Some(6.0) -> (3.0, 50.0), Some(7.0) -> (1.0, 70.0),
+        Some(8.0) -> (1.0, 80.0), None -> (1.0, 90.0), Some(-1.0) -> (9.0, 50.0)))
+
+    model.stats.zip(expected_stats).foreach{
+      case (actual, expected) => assert(actual.equals(expected))
+    }
+
+    val output = model.transform(df_null)
+
+    assert_true(
+      output("output1") === output("expected1") &&
+        output("output2") === output("expected2") &&
+        output("output3") === output("expected3"))
+
+  }
+
+  test("TargetEncoder - unseen null category") {
+
+    val df = spark
+      .createDataFrame(sc.parallelize(data_continuous), schema)
+
+    val encoder = new TargetEncoder()
+      .setLabelCol("label")
+      .setTargetType(TargetEncoder.TARGET_CONTINUOUS)
+      .setHandleInvalid(TargetEncoder.KEEP_INVALID)
+      .setInputCols(Array("input1", "input2", "input3"))
+      .setOutputCols(Array("output1", "output2", "output3"))
+
+    val data_null = Row(null, null, null, 90.0, 50.0, 50.0, 50.0, 57.5, 50.0, 70.0)
+
+    val df_null = spark
+      .createDataFrame(sc.parallelize(data_continuous :+ data_null), schema)
+
+    val model = encoder.fit(df)
+
+    val output = model.transform(df_null)
+
+    assert_true(
+      output("output1") === output("expected1") &&
+        output("output2") === output("expected2") &&
+        output("output3") === output("expected3"))
+
+  }
+
+  test("TargetEncoder - non-indexed categories") {
+
+    val encoder = new TargetEncoder()
+      .setLabelCol("label")
+      .setTargetType(TargetEncoder.TARGET_BINARY)
+      .setInputCols(Array("input1", "input2", "input3"))
+      .setOutputCols(Array("output1", "output2", "output3"))
+
+    val data_noindex = Row(0.toShort, 3, 5.1, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0)
+
+    val df_noindex = spark
+      .createDataFrame(sc.parallelize(data_binary :+ data_noindex), schema)
+
+    val ex = intercept[SparkException] {
+      val model = encoder.fit(df_noindex)
+      print(model.stats)
+    }
+
+    assert(ex.isInstanceOf[SparkException])
+    assert(ex.getMessage.contains(
+      "Values from column input3 must be indices, but got 5.1"))
+
+  }
+
+  test("TargetEncoder - null label") {
```

Comment:
how do it handle `NaN`? treat as a normal value or invalid value?

Reply:
it's an invalid value
` org.apache.spark.SparkException: Values from column input3 must be indices, but got NaN.`

Reply:
- NaN feature => invalid (only accepting indices and null)
- NaN labels => was failing, it's fixed now (observation not considered)

Reply:
> * NaN features => invalid (only accepting indices and null)
> * NaN labels => was failing, it's fixed now (observation not considered)

hi @rebo16v 

https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.TargetEncoder.html#targetencoder

> TargetEncoder considers missing values, such as np.nan or None, as another category and encodes them like any other category. Categories that are not seen during fit are encoded with the target mean, i.e. target_mean_.

It seems scikit-learn's implementation also treat NaN as a valid missing values